src/deluge/model/mod_controllable/mod_controllable_audio.cpp

```diff
@@ -174,6 +194,60 @@ void ModControllableAudio::processFX(StereoSample* buffer, int32_t numSamples, M
 			*postFXVolume = multiply_32x32_rshift32(*postFXVolume, 1518500250) << 1; // Divide by sqrt(2)
 		}
+		else if (modFXType == ModFXType::GRAIN) {
+			if (!grainInitialized && modFXGrainBufferWriteIndex >= 65536) {
+				grainInitialized = true;
+			}
+			*postFXVolume = multiply_32x32_rshift32(*postFXVolume, 1518500250) << 1; // Divide by sqrt(2)
+			// Shift
+			grainShift = 44 * 300; //(kSampleRate / 1000) * 300;
+			// Size
+			grainSize = 44 * ((((unpatchedParams->getValue(Param::Unpatched::MOD_FX_OFFSET) >> 1) + 1073741824) >> 21));
+			grainSize = std::max<int32_t>(440, std::min<int32_t>(35280, grainSize)); //10ms - 800ms
+			// Rate
+			int32_t grainRateRaw =
+			    std::max<int32_t>(0, (std::min<int32_t>(256, ((quickLog(modFXRate) - 364249088) >> 21))));
+			grainRate = ((360 * grainRateRaw >> 8) * grainRateRaw >> 8); //0 - 180hz
+			grainRate = std::max<int32_t>(1, grainRate);
+			grainRate = (kSampleRate << 1) / grainRate;
+			// Preset 0=default
+			grainPitchType = (int8_t)(multiply_32x32_rshift32_rounded(
+			    unpatchedParams->getValue(Param::Unpatched::MOD_FX_FEEDBACK), 5));
+			grainPitchType = std::max<int8_t>(-2, (std::min<int8_t>(2, grainPitchType)));
+			// Temposync
+			if (grainPitchType == 2) {
+				int tempoBPM = (int32_t)(playbackHandler.calculateBPM(currentSong->getTimePerTimerTickFloat()) + 0.5);
+				grainRate = std::max<int32_t>(0, (std::min<int32_t>(256, 256 - grainRateRaw))) << 4; //4096msec
+				grainRate = 44 * grainRate;                              //(kSampleRate*grainRate)/1000;
+				int32_t baseNoteSamples = (kSampleRate * 60 / tempoBPM); //4th
+				if (grainRate < baseNoteSamples) {
+					baseNoteSamples = baseNoteSamples >> 2; //16th
+				}
+				grainRate = std::max<int32_t>(
+				    baseNoteSamples, std::min<int32_t>(baseNoteSamples << 2,
+				                                       (grainRate / baseNoteSamples) * baseNoteSamples)); //Quantize
+				if (grainRate < 2205) {                                                                   //50ms = 20hz
+					grainSize = std::min<int32_t>(grainSize, grainRate << 3) - 1; //16 layers=<<4, 8layers = <<3
+				}
+				bool currentTickCountIsZero = (playbackHandler.getCurrentInternalTickCount() == 0);
+				if (grainLastTickCountIsZero && currentTickCountIsZero == false) { //Start Playback
+					modFXGrainBufferWriteIndex = 0;                                //Reset WriteIndex
+				}
+				grainLastTickCountIsZero = currentTickCountIsZero;
+			}
+			// Rate Adjustment
+			if (grainRate < 882) {                                            //50hz or more
+				grainSize = std::min<int32_t>(grainSize, grainRate << 3) - 1; //16 layers=<<4, 8layers = <<3
+			}
+			// Volume
+			grainVol = modFXDepth - 2147483648;
+			grainVol =
+			    (multiply_32x32_rshift32_rounded(multiply_32x32_rshift32_rounded(grainVol, grainVol), grainVol) << 2)
+			    + 2147483648; //Cubic
+			grainVol = std::max<int32_t>(0, std::min<int32_t>(2147483647, grainVol));
+			grainDryVol =
```

Comment:
It would be nice to have feedback scale with the mix level - grainFeedbackVol = grainVol >> 3; works well in my experimentation. The current feedback makes a pulsing sound with sustained notes and I'd like to be able to have lower mixes without it